cachito/workers/tasks/gomod.py

```diff
@@ -16,45 +17,96 @@
 log = logging.getLogger(__name__)
 
 
+def _find_invalid_gomod_files(bundle_dir, subpaths):
+    """
+    Find all go modules with missing gomod files.
+
+    These files will need to be present in order for the package manager to proceed with
+    fetching the package sources.
+
+    :param RequestBundleDir bundle_dir: the ``RequestBundleDir`` object for the request
+    :param list subpaths: a list of subpaths in the source repository of gomod packages
+    :return: a list containing all non-existing go.mod files across subpaths
+    :rtype: list
+    """
+    invalid_gomod_files = []
+    for subpath in subpaths:
+        bundle_dir_subpath = bundle_dir.app_subpath(subpath)
+        package_gomod_rel_path = bundle_dir_subpath.relpath(bundle_dir_subpath.go_mod_file)
+        log.debug("Testing for go mod file in {}".format(package_gomod_rel_path))
+        if not bundle_dir_subpath.go_mod_file.exists():
+            invalid_gomod_files.append(package_gomod_rel_path)
+
+    return invalid_gomod_files
+
+
 @app.task
-def fetch_gomod_source(request_id, dep_replacements=None):
+def fetch_gomod_source(request_id, dep_replacements=None, package_configs=None):
     """
     Resolve and fetch gomod dependencies for a given request.
 
     :param int request_id: the Cachito request ID this is for
     :param list dep_replacements: dependency replacements with the keys "name" and "version"
+    :param list package_configs: the list of optional package configurations submitted by the user
     :raises CachitoError: if the dependencies could not be retrieved
     """
     config = get_worker_config()
+    if package_configs is None:
+        package_configs = []
+
     bundle_dir = RequestBundleDir(request_id)
+    subpaths = [os.path.normpath(c["path"]) for c in package_configs if c.get("path")]
+
+    if not subpaths:
+        # Default to the root of the application source
+        subpaths = [os.curdir]
+
+    invalid_gomod_files = _find_invalid_gomod_files(bundle_dir, subpaths)
+    if invalid_gomod_files:
+        invalid_files_print = "; ".join(invalid_gomod_files)
+        file_suffix = "s" if len(invalid_gomod_files) > 1 else ""
 
-    if not bundle_dir.go_mod_file.exists():
         if config.cachito_gomod_ignore_missing_gomod_file:
-            log.warning("The go.mod file is missing for request %d", request_id)
+            log.warning("go.mod file%s missing for request at %s", file_suffix, invalid_files_print)
             return
```

Comment:
How should we handle the scenario when only some of the go.mod files are missing? With this, we would not process any of the gomod packages. I would expect Cachito to skip only the packages with missing go.mod and process the ones that do have it.

Reply:
That is a fair point, but I think that the setting to ignore missing gomod files is for backwards compatibility to allow Cachito to just resolve the app directory. If there is only one path, either one is passed or it is the default `.` path. While this situation does not trigger your scenario, it is the design case for resolving.

If a user manually provides multiple paths in the configuration, I would expect that they would intend for all of them to be valid go module paths -- and therefore this behavior is not abnormal. 

Regardless, I agree that something needs to be done here -- either a change in the code or an update of the documentation. Let me know which you think is best!

Reply:
Changes pushed to implement my proposal where it is only supported if one path is supplied. (and documented)